--- a/showiframe.c	2016-02-29 23:47:37.585864637 +0100
+++ b/showiframe.c	2016-02-29 23:47:55.188229540 +0100
@@ -160,11 +160,19 @@
 #endif
 	while(pos <= (s.st_size-4) && !(seq_end_avail = (!iframe[pos] && !iframe[pos+1] && iframe[pos+2] == 1 && iframe[pos+3] == 0xB7)))
 		++pos;
-	if ((iframe[3] >> 4) != 0xE) // no pes header
-		write_all(fd, pes_header, sizeof(pes_header));
-	else
-		iframe[4] = iframe[5] = 0x00;
-	write_all(fd, iframe, s.st_size);
+
+	int count = 7;
+	while(count--){
+		if ((iframe[3] >> 4) != 0xE) { // no pes header
+			write_all(fd, pes_header, sizeof(pes_header));
+			usleep(8000);
+		}
+		else {
+			iframe[4] = iframe[5] = 0x00;
+		}
+		write_all(fd, iframe, s.st_size);
+		usleep(8000);
+	}
 	if (!seq_end_avail)
 		write_all(fd, seq_end, sizeof(seq_end));
 	write_all(fd, stuffing, 8192);
