diff -Naur samba-4.1.12.orig/lib/replace/test/testsuite.c samba-4.1.12/lib/replace/test/testsuite.c
--- samba-4.1.12.orig/lib/replace/test/testsuite.c	2016-01-21 14:47:56.149884449 +0100
+++ samba-4.1.12/lib/replace/test/testsuite.c	2016-01-21 14:48:16.228158572 +0100
@@ -31,9 +31,9 @@
   we include all the system/ include files here so that libreplace tests
   them in the build farm
 */
-#include "system/capability.h"
 #include "system/dir.h"
 #include "system/filesys.h"
+#include "system/capability.h"
 #include "system/glob.h"
 #include "system/iconv.h"
 #include "system/locale.h"
diff -Naur samba-4.1.12.orig/source3/lib/system.c samba-4.1.12/source3/lib/system.c
--- samba-4.1.12.orig/source3/lib/system.c	2016-01-21 23:23:39.434389203 +0100
+++ samba-4.1.12/source3/lib/system.c	2016-01-21 23:24:07.797623370 +0100
@@ -22,9 +22,9 @@
 
 #include "includes.h"
 #include "system/syslog.h"
-#include "system/capability.h"
 #include "system/passwd.h"
 #include "system/filesys.h"
+#include "system/capability.h"
 #include "../lib/util/setid.h"
 
 #ifdef HAVE_SYS_SYSCTL_H
