diff -Naur ldb-1.1.17.orig/lib/replace/test/testsuite.c ldb-1.1.17/lib/replace/test/testsuite.c
--- ldb-1.1.17.orig/lib/replace/test/testsuite.c	2016-01-21 12:48:24.166685071 +0100
+++ ldb-1.1.17/lib/replace/test/testsuite.c	2016-01-21 12:49:07.107936542 +0100
@@ -31,9 +31,9 @@
   we include all the system/ include files here so that libreplace tests
   them in the build farm
 */
-#include "system/capability.h"
 #include "system/dir.h"
 #include "system/filesys.h"
+#include "system/capability.h"
 #include "system/glob.h"
 #include "system/iconv.h"
 #include "system/locale.h"
